grpclog/component.go

```diff
@@ -0,0 +1,289 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package grpclog
+
+import (
+	"fmt"
+	"math"
+	"os"
+	"strconv"
+	"strings"
+
+	"google.golang.org/grpc/internal/grpclog"
+)
+
+// Component is a unique identifier for a component. Created with NewComponent.
+type Component string
+
+// componentData records the settings for a component.
+type componentData struct {
+	verbosity int
+	level     int
+}
+
+const envName = "GRPC_GO_LOG_LEVEL"
+const sentinel = math.MinInt32
+
+// grpc components
+const (
+	CORE       = Component("CORE")
+	ALTS       = Component("ALTS")
+	GRPCLB     = Component("BALANCER_GRPCLB")
+	RLS        = Component("BALANCER_RLS")
+	ROUNDROBIN = Component("BALANCER_ROUNDROBIN")
+	BINARYLOG  = Component("BINARYLOG")
+	CHANNELZ   = Component("CHANNELZ")
+	DNS        = Component("RESOLVER_DNS")
+	TRANSPORT  = Component("TRANSPORT")
+	XDS        = Component("XDS")
+)
```

Comment:
Or make `Component` a (_<IDK what, not singleton? factory? Whatever>_) thing that when it's called the second time with the same string, it returns the component instead of creating a new one.

Each package can just do `var logger = grpclog.Component("core")`. If the component is shared by multiple packages, they will get the same one.

And do we want this function to be only called in `init()`?

Reply:
> Let's define these in the packages they are needed in.

Actually, ignore all the capitalization comments I made earlier.  I still think we should do this.

Reply:
Moved it so they are defined in the packages they are needed. The grpclog package has a cache to strore the components so they are not redefined (is an O(n) operation over the environment variable key-value pairs to create a component).